src/backend/distributed/executor/executor.c

```diff
@@ -698,22 +699,29 @@ TaskListModifiesDatabase(CmdType operation, List *taskList)
 static bool
 DistributedExecutionRequiresRollback(DistributedExecution *execution)
 {
+	CmdType operation = execution->operation;
 	List *taskList = execution->tasksToExecute;
+	int taskCount = list_length(taskList);
 	Task *task = NULL;
 
 	if (MultiShardCommitProtocol == COMMIT_PROTOCOL_BARE)
 	{
 		return false;
 	}
 
-	if (execution->operation == CMD_SELECT)
+	if (taskCount == 0)
 	{
-		return SelectOpensTransactionBlock && IsTransactionBlock();
+		return false;
 	}
 
-	if (taskList == NIL)
+	if (operation == CMD_SELECT)
```

Comment:
wondering if we can avoid explicitly checking for operations (given upcoming CTE changes)

```c
task = (Task *) linitial(taskList);

selectForUpdate = task->relationRowLockList != NIL;
if (selectForUpdate)
{
    return IsTransactionBlock();
}

if (ReadOnlyTask(task))
{
    return SelectOpensTransactionBlock && IsTransactionBlock();
}
```